server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4

```diff
@@ -15,9 +15,59 @@
 lexer grammar CobolLexer;
 import CICSLexer;
 channels{TECHNICAL}
+@header {
+  import java.util.regex.Matcher;
+  import java.util.regex.Pattern;
+  import org.apache.commons.lang3.StringUtils;
+}
+@lexer::members {
+private Pattern titlePattern = Pattern.compile("TITLE\\s*(.*?)\\.?\\s\\n.*", Pattern.CASE_INSENSITIVE);
+    private void isTitlePresent() {
+    String input = _input.getText(Interval.of(_tokenStartCharIndex, _input.index()));
+          if(input!=null) {
+                Matcher matcher = titlePattern.matcher(input);
+                if(matcher.matches()){
+                String matchedTitle = matcher.group(1);
+                if(matchedTitle.isEmpty())
+                reportWrongArguments("lexer.titleCompilerDirective");
+                }
+          }
+        }
+    private void reportWrongArguments(String msg) {
+        ANTLRErrorListener listener = getErrorListenerDispatch();
+        String text = _input.getText(Interval.of(_tokenStartCharIndex, _input.index()));
+	    int stop = _input.index();
+	    if(text.contains("\r")) {
+		   stop--;
+	    }
+        CommonToken lspToken = new CommonTokenFactory().create(_tokenFactorySourcePair, _type, text,
+	         					_channel, _input.index()-text.length()+1,
+	         					stop,
+	        					_tokenStartLine, _tokenStartCharPositionInLine);
+        lspToken.setTokenIndex(_tokenStartCharPositionInLine);
+        lspToken.setText(text.replace("\r","").replace("\n",""));
+        listener.syntaxError(this, lspToken, _tokenStartLine,
+         	         _tokenStartCharPositionInLine, msg, null);
+    }
 
-TITLESTATEMENT : (T I T L E ' '+ .*? NEWLINE) -> skip;
+    private void isLanguageNamePresent() {
+    String input = _input.getText(Interval.of(_tokenStartCharIndex, _input.index()))
+    .replace("\r","").replace("\n","").trim();
+    if(StringUtils.isBlank(input.substring(5, input.length()-1))) {
+    reportWrongArguments("lexer.langMissingEnterDirective");
+    }
+    }
+}
 
+// compiler directive tokens
+TITLESTATEMENT : (T I T L E ' '+ .*? NEWLINE) {isTitlePresent();} -> channel(TECHNICAL);
+
+CONTROL_DIRECTIVE: ASTERISKCHAR (CONTROL | C B L) ((' '| COMMACHAR)
```

Comment:
Any specific reason not defining this (CONTROL_DIRECTIVE) as rule in parser? (want to use benefit of channel ?)

Reply:
Yes, these directives can come anywhere in a source code. something like whitespace. So, we want to throw them in a diff channel/skip. i.e. we don't want to process these tokens.